ui/panels-plugin/src/plugins/time-series-chart/TimeSeriesChartPanel.tsx

```diff
@@ -165,15 +177,17 @@ export function TimeSeriesChartPanel(props: TimeSeriesChartProps) {
 
         // Fallback is unlikely to set unless echarts theme palette in charts theme provider is undefined.
         const fallbackColor =
-          Array.isArray(echartsPalette) && echartsPalette[0]
-            ? (echartsPalette[0] as string)
+          Array.isArray(categoricalPalette) && categoricalPalette[0]
+            ? (categoricalPalette[0] as string) // needed since echarts color property isn't always an array
             : muiTheme.palette.primary.main;
-        // Check which color palette was chosen and get appropriate color.
+
+        // Decide which palette to use based on total series returned.
+        // When series number exceeds number of colors in palette, use generative colors instead.
+        const paletteLength = Array.isArray(categoricalPalette) ? categoricalPalette.length : 4;
         const seriesColor =
-          visual.palette?.kind === 'Categorical'
-            ? getCategoricalPaletteColor(echartsPalette as string[], seriesIndex, fallbackColor)
+          totalSeries <= paletteLength
```

Comment:
What do you think about putting the color-related logic into a separate function/file (ex: `getSeriesColor(queryResults?.data?.series)`)? I think it'll make it easier to follow.

Reply:
Good call, added `getSeriesColor` function and some additional tests [here](https://github.com/perses/perses/blob/5bff142bfe259cdc8679de4e1f1e61ab2ffcdcb4/ui/panels-plugin/src/plugins/time-series-chart/utils/palette-gen.test.ts#L21)!